native/src/main/java/io/ballerina/stdlib/log/MaskedStringBuilder.java

```diff
@@ -0,0 +1,578 @@
+        Map<String, BMap<?, ?>> cached = ANNOTATION_CACHE.get(recordType);
+        if (cached != null) {
+            return cached;
+        }
+
+        // Implement simple cache size management
+        if (ANNOTATION_CACHE.size() >= MAX_CACHE_SIZE) {
+            // Clear half the cache when it gets too large (simple eviction strategy)
+            ANNOTATION_CACHE.entrySet().removeIf(entry -> System.identityHashCode(entry.getKey()) % 2 == 0);
+        }
+
+        Map<String, BMap<?, ?>> annotations = extractFieldAnnotations(recordType);
+        ANNOTATION_CACHE.put(recordType, annotations);
+        return annotations;
+    }
+
+    private String processTableValue(BTable<?, ?> tableValue) {
+        Collection<?> values = tableValue.values();
+        if (values.isEmpty()) {
+            return "[]";
+        }
+
+        int startPos = this.stringBuilder.length();
+        this.stringBuilder.append('[');
+
+        boolean first = true;
+        for (Object row : values) {
+            if (!first) {
+                this.stringBuilder.append(',');
+            }
+            String elementString = buildInternal(row);
+            appendValueToArray(elementString, row);
+            first = false;
+        }
+
+        this.stringBuilder.append(']');
+
+        String result = this.stringBuilder.substring(startPos);
+        this.stringBuilder.setLength(startPos);
+        return result;
+    }
+
+    private String processArrayValue(BArray listValue) {
+        long length = listValue.getLength();
+        if (length == 0) {
+            return "[]";
+        }
+
+        int startPos = this.stringBuilder.length();
+        this.stringBuilder.append('[');
+
+        // Using traditional for loop instead of for-each loop since BArray giving
+        // this error: Cannot read the array length because "<local5>" is null
+        for (long i = 0; i < length; i++) {
+            if (i > 0) {
+                this.stringBuilder.append(',');
+            }
+            Object element = listValue.get(i);
+            String elementString = buildInternal(element);
+            appendValueToArray(elementString, element);
+        }
+
+        this.stringBuilder.append(']');
+
+        String result = this.stringBuilder.substring(startPos);
+        this.stringBuilder.setLength(startPos);
+        return result;
+    }
+
+    private void appendValueToArray(String value, Object originalValue) {
+        if (originalValue instanceof BString) {
+            this.stringBuilder.append('"');
+            appendEscapedString(value);
+            this.stringBuilder.append('"');
+        } else {
+            this.stringBuilder.append(value);
+        }
+    }
+
+    /**
+     * Check if a value is a basic type that doesn't need complex processing.
+     */
+    private static boolean isBasicType(Object value) {
+        return value == null || TypeUtils.getType(value).getTag() <= TypeTags.BOOLEAN_TAG;
```

Comment:
We have to consider the subtypes of string and int here. 

```suggestion
        return value == null || TypeUtils.getType(value).getTag() < TypeTags.NULL_TAG;
```

Reply:
Fixed via [5e45310](https://github.com/ballerina-platform/module-ballerina-log/pull/1343/commits/5e453103126dc42724c847b41e88845e28411134)